File: utils/protoExtendtion.py
# -*- coding: utf-8 -*-
import os,shutil,time

def writeProtoExtendition(pkg,msgAry,clientFile,serveFile):
	t = time.strftime('%Y-%m-%d %H:%M:%S',time.localtime())  
	print(t);
	with open(clientFile,'w') as f:

		f.write("/** \n * @description Auto Generate By protoExtendtion.py\n * @version %s \n * @author suo\n */\n" %t);

		f.write("class ProtoExtendtion {\n");
		f.write("\tpublic static protoMap: Dictionary;\n");
		f.write("\tpublic static init(): void {\n");
		f.write("\t\tthis.protoMap = new Dictionary();\n");
		for item in msgAry:
			f.write("\t\tthis.protoMap.set(\""+ pkg+"."+item +"\","+pkg+"."+item+");\n")
		f.write("\t}\n");
		f.write("}\n");


		f.write("declare namespace "+ pkg +"{\n");
		for item in msgAry:
			f.write("\tinterface "+item+ " {\n")
			f.write("\t\tGetType(): string;\n")
			f.write("\t}\n")
		f.write("}\n");
		for item in msgAry:
			f.write(pkg+ "."+item+".prototype.GetType = function () {\n")
			f.write("\treturn \""+pkg+"."+item +"\";\n")
			f.write("}\n")
		print(u"生成客户端protoExtendition成功")

	with open(serveFile,'w') as f:
		f.write("/** \n * @description Auto Generate By protoExtendtion.py\n * @version %s \n * @author suo\n */\n" %t);
		f.write("import { Cmd } from \"" + r"../../protobuf/common" + "\";\n");
		for item in msgAry:
			f.write(pkg + "." + item + ".prototype['GetType'] = \"" + pkg + "." + item + "\";\n")
		print(u"生成服务器protoExtendition成功")

def openProto(clientFile,serveFile):
	with open(r"..\resource\proto\common.proto", 'r') as f:
		msgAry = [];
		pkg = '';
		list = f.readlines();
		for item in list:
			if item.find('message') == 0:
				msgType = item[len('message'):]
				msgType = msgType.replace('{','')
				msgType = msgType.strip();
				msgAry.append(msgType);
			elif item.find('package') == 0:
				pkgName = item[len('package'):];
				pkgName = pkgName.strip()
				pkg = pkgName[0:-1]
		writeProtoExtendition(pkg,msgAry,clientFile,serveFile);

clientFile = r"..\Egret\src\Extendtion\ProtoExtendtion.ts";
serveFile = r"..\Nodejs\src\Extendtion\ProtoExtendtion.ts"
openProto(clientFile,serveFile);

File: Egret/src/Extendtion/ProtoExtendtion.ts
/** 
 * @description Auto Generate By protoExtendtion.py
 * @version 2019-03-04 10:12:27 
 * @author suo
 */
class ProtoExtendtion {
	public static protoMap: Dictionary;
	public static init(): void {
		this.protoMap = new Dictionary();
		this.protoMap.set("Login_C",Cmd.Login_C);
		this.protoMap.set("PlayerInfo_S",Cmd.PlayerInfo_S);
		this.protoMap.set("ItemInfo_CS",Cmd.ItemInfo_CS);
		this.protoMap.set("ItemUpdate_CS",Cmd.ItemUpdate_CS);
		this.protoMap.set("Heartbeat_CS",Cmd.Heartbeat_CS);
		this.protoMap.set("TaskUpdate_CS",Cmd.TaskUpdate_CS);
		this.protoMap.set("RefreshTask_C",Cmd.RefreshTask_C);
		this.protoMap.set("GetTaskAward_C",Cmd.GetTaskAward_C);
		this.protoMap.set("SameUidLogin_S",Cmd.SameUidLogin_S);
		this.protoMap.set("ServeTips_S",Cmd.ServeTips_S);
	}
}
declare namespace Cmd{
	interface Login_C {
		GetType(): string;
	}
	interface PlayerInfo_S {
		GetType(): string;
	}
	interface ItemInfo_CS {
		GetType(): string;
	}
	interface ItemUpdate_CS {
		GetType(): string;
	}
	interface Heartbeat_CS {
		GetType(): string;
	}
	interface TaskUpdate_CS {
		GetType(): string;
	}
	interface RefreshTask_C {
		GetType(): string;
	}
	interface GetTaskAward_C {
		GetType(): string;
	}
	interface SameUidLogin_S {
		GetType(): string;
	}
	interface ServeTips_S {
		GetType(): string;
	}
}
Cmd.Login_C.prototype.GetType = function () {
	return "Cmd.Login_C";
}
Cmd.PlayerInfo_S.prototype.GetType = function () {
	return "Cmd.PlayerInfo_S";
}
Cmd.ItemInfo_CS.prototype.GetType = function () {
	return "Cmd.ItemInfo_CS";
}
Cmd.ItemUpdate_CS.prototype.GetType = function () {
	return "Cmd.ItemUpdate_CS";
}
Cmd.Heartbeat_CS.prototype.GetType = function () {
	return "Cmd.Heartbeat_CS";
}
Cmd.TaskUpdate_CS.prototype.GetType = function () {
	return "Cmd.TaskUpdate_CS";
}
Cmd.RefreshTask_C.prototype.GetType = function () {
	return "Cmd.RefreshTask_C";
}
Cmd.GetTaskAward_C.prototype.GetType = function () {
	return "Cmd.GetTaskAward_C";
}
Cmd.SameUidLogin_S.prototype.GetType = function () {
	return "Cmd.SameUidLogin_S";
}
Cmd.ServeTips_S.prototype.GetType = function () {
	return "Cmd.ServeTips_S";
}
